examples/vizdoc/html/heatmap.html

```diff
@@ -0,0 +1,193 @@
+    //return f;
+}
+
+calcval();
+
+var svg = d3.select("svg"),
+    width = +svg.attr("width"),
+    height = +svg.attr("height");
+
+// needed a lot of fine-tuning still not ideal
+var thresholds = d3.range(0, 61)
+    .map(function(p) { return Math.pow(1.3, p); });
+
+
+//  var thresholds = [0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 100000, 110000, 120000, 40000, 21 ** 2]
+//  var thresholds = d3.range(0, 12)
+//     .map(function(p) { return p * 1000; }).push(40000, 21 ** 2);
+
+var contours = d3.contours()
+    .size([n, m])
+    .thresholds(thresholds);
+
+var color = d3.scaleLog()
+    .domain(d3.extent(thresholds.map(function(p) {return p + 1800})))
+    .interpolate(function() { return d3.interpolateYlGnBu; });
+
+// make contours
+svg.selectAll("path")
+  .data(contours(values))
+  .enter().append("path")
+    .attr("d", d3.geoPath(d3.geoIdentity().scale(width / n)))
+    .attr("fill", function(d) { return color(d.value); });
+
+// add circles
+svg.append("circle")
+  .attr("cx", circ1[0]).attr("cy", circ1[1]).attr("r", 20).attr("id", "circ1").style("fill", "blue");
+svg.append("circle")
+  .attr("cx", circ2[0]).attr("cy", circ2[1]).attr("r", 20).attr("id", "circ2").style("fill", "green");
+
+// update on input change
+d3.select("#circ1x").on("input", function() {
+  update(this.id, +this.value);
+});
+d3.select("#circ1y").on("input", function() {
+  update(this.id, +this.value);
+});
+d3.select("#circ2x").on("input", function() {
+  update(this.id, +this.value);
+});
+d3.select("#circ2y").on("input", function() {
+  update(this.id, +this.value);
+});
+
+// Initial starting positions of the circles
+update("circ1x", circ1[0]);
+update("circ1y", circ1[1]);
+update("circ2x", circ2[0]);
+update("circ2y", circ2[1]);
+
+function update(id, val) {
+    console.log("update", id, val);
+
+  // adjust the positions on the range slider
+  // not elegant, should be refactored tbh
+  if (id == "circ1x") {
+    d3.select("#circ1x-value").text(val);
+    d3.select("#circ1x").property("value", val);
+    d3.select("#circ1").attr(
+        "cx", val
+    );
+    circ1[0] = val;
+  }
+  else if (id == "circ1y") {
+    d3.select("#circ1y-value").text(val);
+    d3.select("#circ1y").property("value", val);
+    d3.select("#circ1").attr(
+        "cy", val
+    );
+    circ1[1] = val;
+  }
+  else if (id == "circ2x") {
+    d3.select("#circ2x-value").text(val);
+    d3.select("#circ2x").property("value", val);
+    d3.select("#circ2").attr(
+        "cx", val
+    );
+    circ2[0] = val;
+  }
+  else if (id == "circ2y") {
+    d3.select("#circ2y-value").text(val);
+    d3.select("#circ2y").property("value", val);
+    d3.select("#circ2").attr(
+        "cy", val
+    );
+    circ2[1] = val;
+  }
+  calcval();
+
+    // Note: If your code isn't working, try googling more specific search terms (I got here by searching "d3 update selection svg", since it seemed like the update selection for existing elements in the svg wasn't working)
```

Comment:
You can delete my old comments in this file @strout18

Reply:
Addressed in 2b9315a4a27f35fb751ac317f15cdbf9a18ec5e4